### PR TITLE
Fix issue #28: Unintuitive adapter syntax

### PR DESCRIPTION
**Pull request by grzes314**

### CODE REVIEW

File: steps/base.py [hunk 1]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
-                    func = take_first_inputs
-                else:
-                    raise ValueError('wrong mapping specified')
-
-                raw_inputs = [step_inputs[step_name][step_var] for step_name, step_var in step_mapping]
-                adapted_steps[adapted_name] = func(raw_inputs)
+        for adapted_name, recipe in self.adapter.items():
+            try:
+                adapted_steps[adapted_name] = self._adapt_one_name(step_inputs, recipe)
+            except (KeyError, ValueError) as ex:
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 I think the usual naming is simply e

File: steps/base.py [hunk 2]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
+                raise StepsError("Error while adapting '{}'".format(adapted_name)) from ex
+
         return adapted_steps
 
     def unpack(self, step_inputs):
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 again I know it is on me but regardles I think unpack should be private

File: steps/base.py [hunk 3]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
         logger.info('step {} unpacking inputs'.format(self.name))
         unpacked_steps = {}
+        key_to_step_names = defaultdict(list)
         for step_name, step_dict in step_inputs.items():
-            unpacked_steps = {**unpacked_steps, **step_dict}
-        return unpacked_steps
+            unpacked_steps.update(step_dict)
+            for key in step_dict.keys():
+                key_to_step_names[key].append(step_name)
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 I have to say I use key_to_step_names.setdefault(key, []).append(step_name) instead of defaultdict(list) but it looks way cleaner here.

File: steps/base.py [hunk 4]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
         all_steps = self._get_steps(all_steps)
         return all_steps
 
+    def _adapt_one_name(self, step_inputs, recipe):
+        if isinstance(recipe, tuple) and len(recipe) == 2 and not callable(recipe[1]):
+            input_name, key = recipe
+            return self._extract_one_item(step_inputs, input_name, key)
+
+        if isinstance(recipe, tuple) and len(recipe) == 2 and callable(recipe[1]):
+            lst = recipe[0]
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 since we assure that it is a tuple of len 2 then simply lst, fun = recipe will do

File: steps/base.py [hunk 5]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
+            fun = recipe[1]
+        elif isinstance(recipe, list):
+            lst = recipe
+            fun = lambda x: x
+        else:
+            raise ValueError("Invalid adapting recipe: '{}'".format(recipe))
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 we followed the msg = ''.format first then pass it to Exception/Error I would keep that convention

File: steps/base.py [hunk 6]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
+                     for input_name, key in lst]
+        return fun(extracted)
+
+    def _extract_one_item(self, step_inputs, input_name, key):
+        try:
+            input_dict = step_inputs[input_name]
+            try:
+                return input_dict[key]
+            except KeyError:
+                raise StepsError("Step '{}' didn't have '{}' in its output."
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 same first msg then pass it to StepsError

File: steps/base.py [hunk 7]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
+                                 .format(input_name, key))
+        except KeyError:
+            raise StepsError("Step '{}' doesn't have '{}' as its input step."
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 same first msg then pass it to StepsError

File: steps/base.py [hunk 8]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
+            fun = lambda x: x
+        else:
+            raise ValueError("Invalid adapting recipe: '{}'".format(recipe))
+
+        extracted = [self._extract_one_item(step_inputs, input_name, key)
+                     for input_name, key in lst]
+        return fun(extracted)
+
+    def _extract_one_item(self, step_inputs, input_name, key):
+        try:
```
Comment (jakubczakon) [inline steps/base.py]: @grzes314 maybe insead of try catching input_dict = step_inputs.get(input_name) and then if input_dict is not None else raise. I think there is a lot of try_catching here and maybe not all of it necessary. On the other hand user is breaching the contract when she passes incorrect step_inputs so it definitely should raise so why not try catch.

File: tests/steps_test_case.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import logging
```
Comment (jakubczakon) [inline tests/steps_test_case.py]: @grzes314 It is official we are testing! Happy as a clam.

File: tests/test_base.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
+import unittest
```
Comment (jakubczakon) [inline tests/test_base.py]: @grzes314 I am only thinking whether to use unittest or pytest. pytest is generally easier for non dev people  Do you feel strongly about unittest ?

Comment (grzes314) [reply tests/test_base.py]: I think I chose unittest because of its similarity to JUnit which I used before. But I don't insist. I can rewrite these tests in pytest.

File: steps/base.py (hunk 8 above)
Comment (grzes314) [reply steps/base.py]: This is a question of whether to use [EAFP](https://docs.python.org/3.6/glossary.html#term-eafp) style or [LBYL](https://docs.python.org/3.6/glossary.html#term-lbyl) style. I also find LBYL more readable, but AFAIK EAFP is preferred in Python.

File: tests/test_base.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        )
+        res = step.fit_transform(self.data)
+        self.assertTrue(np.array_equal(res['X'], self.data['input_1']['features']))
+
+    def test_adapter_recipe_with_list(self):
+        step = Step(
+            name='test_adapter_recipe_with_list',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': [],
```
Comment (mromaniukcdl) [inline tests/test_base.py]: What's this `'X': []` ?

File: tests/test_base.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+            cache_dirpath=CACHE_DIRPATH
+        )
+        res = step.fit_transform(self.data)
+        for i, key in enumerate(('X', 'Y', 'Z')):
+            self.assertIsInstance(res[key], list)
+            self.assertEqual(len(res[key]), i)
+
+        self.assertEqual(res['X'], [])
+
+        self.assertTrue(np.array_equal(res['Y'][0], self.data['input_1']['features']))
```
Comment (mromaniukcdl) [inline tests/test_base.py]: Why does `res['Y']` have to be further indexed with `[0]`?

File: tests/test_base.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+        self.assertTrue(np.array_equal(res['Z'][0], self.data['input_1']['features']))
+        self.assertTrue(np.array_equal(res['Z'][1], self.data['input_2']['extra_features']))
+
+    def test_adapter_recipe_with_list_and_function(self):
+        step = Step(
+            name='test_adapter_recipe_with_list_and_function',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': ([('input_1', 'features'),
```
Comment (mromaniukcdl) [inline tests/test_base.py]: Personally I think this kind of adapter is hard to understand. My preference would be that if you want to concatenate features from different input steps then you should just create a concatenation step.

File: tests/test_base.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        )
+        with self.assertRaises(StepsError):
+            step.fit_transform(self.data)
+
+    def test_adapter_creates_defined_keys(self):
+        step = Step(
+            name='test_adapter_creates_defined_keys',
+            transformer=Dummy(),
+            input_data=['input_1', 'input_2'],
+            adapter={'X': [('input_1', 'features')],
```
Comment (mromaniukcdl) [inline tests/test_base.py]: Wasn't the reason for making changes to get rid of the square brackets here?

File: steps/base.py [hunk 15]
```diff
@@ -176,35 +176,69 @@ def _cached_transform(self, step_inputs):
...
         return adapted_steps
 
     def unpack(self, step_inputs):
         logger.info('step {} unpacking inputs'.format(self.name))
         unpacked_steps = {}
+        key_to_step_names = defaultdict(list)
         for step_name, step_dict in step_inputs.items():
-            unpacked_steps = {**unpacked_steps, **step_dict}
-        return unpacked_steps
+            unpacked_steps.update(step_dict)
```
[the remark this answers is not in the document]
Comment (grzes314) [reply steps/base.py]: I check if this is a case a few lines further. If yes, an exception is thrown and `unpacked_steps` is discarded.

File: tests/test_base.py (hunk 11 above)
Comment (grzes314) [reply tests/test_base.py]: List of values in input steps to extract is empty, so `X` will be an empty list.

File: tests/test_base.py (hunk 12 above)
Comment (grzes314) [reply tests/test_base.py]: `recipe` for `Y` is a list, so `res['Y']` is a list.

File: tests/test_base.py (hunk 14 above)
Comment (grzes314) [reply tests/test_base.py]: If you do `'X': ('input_1', 'features')`, then `X` will point to the same object as `features` in `input_1`.
If you do `'X': [('input_1', 'features')]`, then `X` will be a singleton list containing those `features`.

File: tests/test_base.py (hunk 13 above)
Comment (grzes314) [reply tests/test_base.py]: First, let me notice that the syntax for this case hasn't changed :)

In this PR I executed my proposition described in issue #28 thread. My idea was to retain as much of existing behaviour as possible, while:
1. Simplifying syntax for the simplest and most popular case (removing brackets if you just want one item).
2. Making the syntax more consistent (recipe containing a list yields a list of objects).

If you have ideas for different approach to adapting inputs present them in issue #28 :)

**Comment by grzes314** [comment]
@jakubczakon I took into account most of your suggestions. Most notably, I rewrote the tests in pytest.

**Comment by grzes314** [comment]
Following our recent discussion I introduced a few important changes in this PR:
- I moved adapting logic to a separate class.
- I removed option of adding a function to adapter.
- It is now possible to assign an arbitrary object to an adapted name.
- Unfortunately, it means that to let the adapter know we want to extract an element from the inputs we have to precede (input, key) pair with letter E, e.g. `'y': E('input_train', 'labels')`. If it was just `'y': ('input_train', 'labels')` there would be no telling if 'y' should be a constant tuple consisting of two strings, or if we should extract something from inputs results.

**The "adapter syntax" can now be (almost) any Python literal** and the adapter preserves structure of that literal's value with all `E(input, key)` replaced by elements extracted from the inputs.
